### ISSUE TITLE
tf.nn.conv2d_transpose why output_shape is needed?

### ISSUE THREAD

**Issue by ziky90** [tensorflow/tensorflow#2118]
I am currently working with `tf.nn.conv2d_transpose`. From the past I am used to Caffe deconvolution layer and `tf.nn.conv2d_transpose` is kind of TensorFlow equivalent to it.

My question here is if someone could point me to detailed behaviour/documenatation of `tf.nn.conv2d_transpose`.
Particularly I am confused by `output_shape` parameter. My question is, why there is needed the `output_shape` parameter. Isn't the output shape directly come from the conv2d_transpose operation?
Based on how I understand it from Caffe and from [here](http://cs231n.stanford.edu/slides/winter1516_lecture13.pdf), computed as:
```
h = ((len(value[1]) - 1) * stride_h) + kernel_h - 2 * pad_h
w = ((len(value[2]) - 1) * stride_w) + kernel_w - 2 * pad_w
```

What happens when I set `output_shape` smaller than `h` and `w`? Is the new layer being cropped?

What happens when I set it higher? Is there being created padding with only zeros?

**Comment by girving** [comment]
`output_shape` is needed because the shape of the output can't necessarily be computed from the shape of the input, specifically if the output is smaller than the filter and we're using `VALID` padding so the input is an empty image.  However, this degenerate case is unimportant most of the time, so it'd be reasonable to make the Python wrapper compute `output_shape` automatically if it isn't set.

**Comment by girving** [comment]
Ah, also the shape is ambiguous in cases with `stride > 1`.

**Comment by ziky90** [comment]
@girving I am still not sure why shape is ambiguous with `stride > 1`? Is it only for `padding=VALID` and `stride>1`?

Giving particular example:
For example let's say that I have `stride=32`, `kernel=64` and `padding=SAME` and `value_w=value_h=16`. Then `output_shape` must be [b, 544, 544, c]. I don't see how I could get different `output_shape`.

**Comment by girving** [comment]
`conv2d_transpose` is also used as the gradient of `conv2d`, so it needs to be able to spit out any shape that can be an input to `conv2d` for a given output shape of `conv2d`.  The formulas for the shape of the output of `conv2d` are

    output = (input - filter + stride) // stride  # VALID
    output = (input + stride - 1) // stride  # SAME

If `stride > 1`, neither of these are invertible.  However, I think it's an excellent idea to give `output_shape` a default value for `conv2d_transpose` in Python.

**Comment by amolchanov86** [comment]
Hi everyone,
I have a question. How can I force certain dimensionality of the output of the conv2d_transpose layer ? My problem is that I use it for upsampling and I want to match the dimensionality of my labels and the output of the NN, for example if I have a feature map as Bx25x40xC how can I make it Bx100x160xC (i.e. upsample exactly 4x times) ?

**Comment by girving** [comment]
@amolchanov86 Questions like that should go on StackOverflow; Github issues are for bugs and feature requests.

**Comment by amolchanov86** [comment]
Well, it would be actually great to have such feature, since the conv2d_inverse is often used for the goal of upsampling anyway and one usually have a standard target size :) But, sure, you are right.

**Comment by girving** [comment]
I'm going to mark this as contributions welcome since there are some subtleties that I don't have time to investigate at the moment.  I'd be happy to review PRs if anyone wants to tackle this.  The subtlety is that  I'm not sure what the ideal default should be.

**Comment by zhangqianhui** [comment]
@girving  I want to know the details of transpose_conv, Do you have more books or links to introduce the code of transpose_conv ?

**Comment by tongxyh** [comment]
i get the problem that why the B(batch_size) is fixed during both train and test. if i change the batch size during test, if will throw error.

**Comment by ncullen93** [comment]
What is the status of this issue?

**Comment by ncullen93** [comment]
I made this function to calculate output shape, assuming stride==1. I'm a little unclear if stride > 1. 
Just something to kickstart this conversation again.

def calculate_output_shape(in_layer, n_kernel, kernel_size, border_mode='same'):
	"""
	Always assumes stride=1
	"""
	in_shape = in_layer.get_shape()
	out_shape = [s for s in in_shape] # copy
	out_shape[-1] = n_kernel # always true
	if border_mode=='same':
		out_shape[1] = in_shape[1]
		out_shape[2] = in_shape[2]
	elif border_mode == 'valid':
		out_shape[1] = in_shape[1]+kernel_size - 1
		out_shape[2] = in_shape[2]+kernel_size - 1
	return out_shape

**Comment by girving** [comment]
@ncullen93 I believe if `stride > 1` you should pick the minimum consistent size, but it would take some more thought to make sure.

**Comment by paolorota** [comment]
one further question, why in the documentation the `output_shape` is referred as a 1D Tensor when it asks for a 4D Tensor instead?

**Comment by girving** [comment]
The *shape* of a tensor is itself a 1-D tensor.  For example, the shape of a 4-D tensor is a 1-D tensor with 4 elements.  This code may make the situation clearer:

    >>> np.array(np.zeros((2,3,4,5)).shape).shape
    (4,)

**Comment by paolorota** [comment]
all clear, sorry for the stupid question! :(

**Comment by 3rd3** [comment]
> so it needs to be able to spit out any shape that can be an input to conv2d for a given output shape of conv2d

@girving Why can't these shapes be inferred from the inputs? Could you provide a concrete example?

**Comment by girving** [comment]
@3rd3 Consider a forward convolution filter bigger than the input image, or a stride bigger than the input image, both with `padding='VALID'`.  Then the output will have size 0 or 1, respectively, independent of the input image.  Running this backwards, if we do a `conv2d_transpose` with input size 0 or 1, it could produce many possible output sizes.

**Comment by boluoweifenda** [comment]
In tensorflow.slim implementation of conv_transpose, I find that output_shape can be calculated by this:
from tensorflow.python.layers import utils
out_H = utils.deconv_output_length(in_H, ksize, padding, stride)
out_W = utils.deconv_output_length(in_W, ksize, padding, stride)

**Comment by LucasMahieu** [comment]
I would say that you can invert 'input' and 'output' in the expression of conv2d output shapes:
for conv2d: 
```
output = (input - filter + stride) // stride  # VALID
output = (input + stride - 1) // stride  # SAME
```

so I would say that for conv2d_transpose:
```
output = input * stride + filter - stride  # VALID
output = input * stride - stride + 1  # SAME 
```

What do you think about that ?

**Comment by philipp-fischer** [comment]
It is pretty simple to show that it is not invertible due to the integer division by stride.
Let's consider the formula for conv2d with padding SAME:
`output = (input + stride - 1) // stride`
Let's say stride is 2 (a common case), then for an input size of 4 we get output size 2.
But for input size 3 we also get an output size of 2. 
So **not invertible**.
Hence we need to specify the output size of the conv2d_transpose layer explicitly

**Comment by gauss-clb** [comment]
@girving 
deconv_output_length(input_length, filter_size, padding, stride) is defined [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/utils.py#L159)

When padding is valid, input_length += max(filter_size - stride, 0), why to use `max` function?

For conv2d:
```
output = (input - filter + stride) // stride  # VALID
output = (input + stride - 1) // stride  # SAME
```

For conv2d_transpose:
```
output = input * stride + filter - stride  # VALID
output = input * stride - stride + 1  # SAME 
```

Even when filter_size is less than stride, I think output is also `input * stride + filter - stride` rather `input * stride`, so why to use `max`?

**Comment by hnyz979** [comment]
@tongxyh I also get this error. If I use output_shape=[input_.get_shape()[0].value, 32, 32, 256], it throws an error "Failed to convert object of type <class 'list'> to Tensor. Contents: [None, 32, 32, 256]. Consider casting elements to a supported type" when building the computation graph.
However, If I use output_shape=[-1, 32, 32, 256], it would be OK when building the graph, but when the session is ran, there also gets an error...

**Comment by Datoclement** [comment]
While the integer division is not invertible, we have no information about the pixels that "belong" to the residual part. So, the consideration of these extra pixels leads us to padding the output tensor. Am I correct?
(Not sure if this question/thought should be posted here.)

e.g.,

To think about this, I use "input" to denote the input of the convolution (non transposed) and "output" its output. Since we are only interested in shapes, output -> conv2d_transposed -> input.

Given input of shape [7, 7], kernel of shape [3, 3]. With valid padding, the output before striding is [5, 5]. If the stride is [3, 3] then the final output is of shape [2, 2].

The same setting with input of shape [6, 6] gives the same output shape [2, 2], where we find the invertibility of integer division.

However, we actually do not have any information about the extra pixels in the first case. (extra pixels in [7, 7] compared to [6, 6]). So, when we want to conv2d_transpose back to the input image [7, 7], we can do nothing but zero paddings. (Maybe translated by a bias later?)

**Comment by guoyejun** [comment]
I see it is not invertible, and the desired output shape for conv2d_transposed is necessary. 

But it becomes a common issue when trying to integrate the model in a c program and execute the model via Tensorflow C API, because the model (with conv2d_transposed layer) is tied with the desired output shape which is usually relative to the input image size. And we have to prepare many model files for different possible input image size, it is not feasible.

Actually, for many cases, the model (for example, with encoder and decoder) is designed to be independent with the input image size, it is expected that a single model file can be used for all kinds of image sizes.

So, can we add an option for this layer to not set output_shape explicitly?
- add a new parameter remainders, and so:
output = input * stride + remainder + filter - stride  # VALID
output = input * stride + remainder - stride + 1  # SAME 

- if the output_shape is none (or even add a new parameter flag), just do:
output = input * stride + filter - stride  # VALID
output = input * stride - stride + 1  # SAME

**Comment by dianyuanrenzhe** [comment]
> i get the problem that why the B(batch_size) is fixed during both train and test. if i change the batch size during test, if will throw error.

**Comment by mohantym** [comment]
Hi @ziky90! We are checking to see if you still need help this feature request , have you tried latest stable version 2.6 yet as many features have been added in latest versions.
